### ISSUE TITLE
py-evm tests will fail with current master branch of py_ecc

### ISSUE THREAD

**Issue by ChihChengLiang** [ethereum/py_ecc#59]
* py-evm Version: master branch at d6adf9039647b6fc39b9d79725e9331008a5c438
* OS: osx
* Python Version (python --version): Python 3.7.0
* Environment (output of `pip freeze`):

### What is wrong?

Please include information like:

* full output of the error you received
* what command you ran: `(in py-evm)$ pytest tests/json-fixtures/test_blockchain.py::test_blockchain_fixtures[.../py-evm/fixtures/BlockchainTests/GeneralStateTests/stCreate2/create2callPrecompiles_d5g0v0.json:create2callPrecompiles_d5g0v0_Constantinople:Constantinople] ` replace ... with your absolute path
* the code that caused the failure: `py_ecc.optimized_bn128.optimized_field_elements.FQ`

```                                                                                                                     

tests/json-fixtures/test_blockchain.py F                                                                                             [100%]

================================================================= FAILURES =================================================================
 test_blockchain_fixtures[py-evm/fixtures/BlockchainTests/GeneralStateTests/stCreate2/create2callPrecompiles_d5g0v0.json:create2callPrecompiles_d5g0v0_Constantinople:Constantinople] 

fixture_data = ('py-evm/fixtures/BlockchainTests/GeneralStateTests/stCreate2/create2callPrecompiles_d5g0v0.json', 'create2callPrecompiles_d5g0v0_Constantinople', 'Constantinople')
fixture = {'blocks': [{'blockHeader': {'bloom': 0, 'coinbase': b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba', '... 'lastblockhash': b'\xd3\x92\r\x904,\x0f\t\xaex$f/\x964\xd7\x8d\x8d\xfb\xf95\xdfK\x0f\x04\n\x81\xa5\x0c\xa2g\x95', ...}

    def test_blockchain_fixtures(fixture_data, fixture):
        try:
            chain = new_chain_from_fixture(fixture)
        except ValueError as e:
            raise AssertionError("could not load chain for {}".format((fixture_data,))) from e
    
        genesis_params = genesis_params_from_fixture(fixture)
        expected_genesis_header = BlockHeader(**genesis_params)
    
        # TODO: find out if this is supposed to pass?
        # if 'genesisRLP' in fixture:
        #     assert rlp.encode(genesis_header) == fixture['genesisRLP']
    
        genesis_block = chain.get_canonical_block_by_number(0)
        genesis_header = genesis_block.header
    
        assert_imported_genesis_header_unchanged(expected_genesis_header, genesis_header)
    
        # 1 - mine the genesis block
        # 2 - loop over blocks:
        #     - apply transactions
        #     - mine block
        # 3 - diff resulting state with expected state
        # 4 - check that all previous blocks were valid
    
        mined_blocks = list()
        for block_fixture in fixture['blocks']:
            should_be_good_block = 'blockHeader' in block_fixture
    
            if 'rlp_error' in block_fixture:
                assert not should_be_good_block
                continue
    
            if should_be_good_block:
                (block, mined_block, block_rlp) = apply_fixture_block_to_chain(
                    block_fixture,
                    chain,
>                   perform_validation=False  # we manually validate below
                )

block_fixture = {'blockHeader': {'bloom': 0, 'coinbase': b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba', 'difficulty':...x00a\x01\x00`\x00`\x00`\x06b\t'\xc0\xf1`\x02U`\x00Q`\x00U\x00", 'gasLimit': 15000000, 'gasPrice': 1, 'nonce': 0, ...}]}
chain      = <abc.ChainFromFixture object at 0x10d834278>
expected_genesis_header = <BlockHeader #0 8a27153c>
fixture    = {'blocks': [{'blockHeader': {'bloom': 0, 'coinbase': b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba', '... 'lastblockhash': b'\xd3\x92\r\x904,\x0f\t\xaex$f/\x964\xd7\x8d\x8d\xfb\xf95\xdfK\x0f\x04\n\x81\xa5\x0c\xa2g\x95', ...}
fixture_data = ('py-evm/fixtures/BlockchainTests/GeneralStateTests/stCreate2/create2callPrecompiles_d5g0v0.json', 'create2callPrecompiles_d5g0v0_Constantinople', 'Constantinople')
genesis_block = <ConstantinopleBlock(#Block #0)>
genesis_header = <BlockHeader #0 8a27153c>
genesis_params = {'block_number': 0, 'bloom': 0, 'coinbase': b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba', 'difficulty': 131072, ...}
mined_blocks = []
should_be_good_block = True

tests/json-fixtures/test_blockchain.py:269: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
eth/tools/fixtures/helpers.py:232: in apply_fixture_block_to_chain
    mined_block, _, _ = chain.import_block(block, perform_validation=perform_validation)
eth/chains/base.py:726: in import_block
    imported_block = self.get_vm(base_header_for_import).import_block(block)
eth/vm/base.py:539: in import_block
    new_header, receipts, _ = self.apply_all_transactions(block.transactions, self.block.header)
eth/vm/base.py:492: in apply_all_transactions
    transaction,
eth/vm/base.py:408: in apply_transaction
    state_root, computation = self.state.apply_transaction(transaction)
eth/vm/state.py:256: in apply_transaction
    computation = self.execute_transaction(transaction)
eth/vm/forks/frontier/state.py:201: in execute_transaction
    return executor(transaction)
eth/vm/state.py:307: in __call__
    computation = self.build_computation(message, valid_transaction)
eth/vm/forks/frontier/state.py:133: in build_computation
    transaction_context,
eth/vm/forks/spurious_dragon/computation.py:33: in apply_create_message
    computation = self.apply_message()
eth/vm/forks/frontier/computation.py:76: in apply_message
    self.transaction_context,
eth/vm/computation.py:583: in apply_computation
    opcode_fn(computation=computation)
eth/vm/logic/system.py:215: in __call__
    return super().__call__(computation)
eth/vm/logic/system.py:193: in __call__
    self.apply_create_message(computation, child_msg)
eth/vm/logic/system.py:252: in apply_create_message
    child_computation = computation.apply_child_computation(child_msg)
eth/vm/computation.py:395: in apply_child_computation
    child_computation = self.generate_child_computation(child_msg)
eth/vm/computation.py:404: in generate_child_computation
    self.transaction_context,
eth/vm/forks/spurious_dragon/computation.py:33: in apply_create_message
    computation = self.apply_message()
eth/vm/forks/frontier/computation.py:76: in apply_message
    self.transaction_context,
eth/vm/computation.py:583: in apply_computation
    opcode_fn(computation=computation)
eth/vm/logic/call.py:136: in __call__
    child_computation = computation.apply_child_computation(child_msg)
eth/vm/computation.py:395: in apply_child_computation
    child_computation = self.generate_child_computation(child_msg)
eth/vm/computation.py:410: in generate_child_computation
    self.transaction_context,
eth/vm/forks/frontier/computation.py:76: in apply_message
    self.transaction_context,
eth/vm/computation.py:569: in apply_computation
    computation.precompiles[message.code_address](computation)
eth/precompiles/ecadd.py:35: in ecadd
    result = _ecadd(computation.msg.data_as_bytes)
eth/precompiles/ecadd.py:59: in _ecadd
    p1 = validate_point(x1, y1)
eth/_utils/bn128.py:26: in validate_point
    if not bn128.is_on_curve(p1, bn128.b):
../py_ecc/py_ecc/optimized_bn128/optimized_curve.py:69: in is_on_curve
    return y**2 * z - x**3 == b * z**3
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 3, other = 1

    def __mul__(self, other: IntOrFQ) -> "FQ":
        if isinstance(other, FQ):
            on = other.n
        elif isinstance(other, int):
            on = other
        else:
            raise TypeError(
                "Expected an int or FQ object, but got object of type {}"
>               .format(type(other))
            )
E           TypeError: Expected an int or FQ object, but got object of type <class 'py_ecc.optimized_bn128.optimized_field_elements.FQ'>

other      = 1
self       = 3

../py_ecc/py_ecc/fields/optimized_field_elements.py:67: TypeError
========================================================= 1 failed in 8.43 seconds =========================================================
```

### How can it be fixed

- might need a separate branch for the release

**Comment by ChihChengLiang** [comment]
cc @pipermerriam @carver @Bhargavasomu

**Comment by Bhargavasomu** [comment]
@ChihChengLiang the fix for this is here at https://github.com/ethereum/py-evm/blob/d6adf9039647b6fc39b9d79725e9331008a5c438/eth/_utils/bn128.py#L17.

Here if we replace this with
```
FQ = optimized_bn128_FQ
```
then this test would work.

This refactoring would make a breaking change in probably a lot of repositories using this library, but it is maybe for the best, as it provides better structure.

**Comment by carver** [comment]
AFAIK, master is not slated for a major version bump, so we need to:
1. Add a new test to cover this case
2. Fix the API so it stays backwards compatible

**Comment by ChihChengLiang** [comment]
@Bhargavasomu The tests are happy with the fix, but not mypy. Add a draft PR here https://github.com/ethereum/py-evm/pull/1731 for the proposed diff.
Despite the issues we encountered, I totally support the refactor.

@carver would propose porting _ecpairing from eth/precompiles/ecpairing.py to py_ecc for better maintainability and compatibility.

To provide more context, my goal is to release BLS features #52 as soon as possible. It is required by the test generator update (https://github.com/ethereum/eth2.0-test-generators/issues/22) and Trinity update (https://github.com/ethereum/trinity/issues/329).

Here are some proposed options:

### 1. Merge #60 
  - pros: no effect in py-evm and extremely fast release. 
  - cons: Negative effect for the py_ecc refactor.
 (I'm in favor of this option)
### 2. In another py_ecc branch, release the BLS features without py_ecc refactor.
  - pros: no effect in py-evm, no effect in py_ecc refactor, and fast release.
  - cons: need to merge the branches together in the future.
### 3. Lock py_ecc version in py-evm at py_ecc==1.4.8 and release BLS in 1.4.9. Then simultaneously fix tests in py-evm and unlock again in a PR.
  - pros: no effect in py_ecc refactor and fast release.
  - cons: long merge in Trinity.

I'm in favor of option 1

**Comment by Bhargavasomu** [comment]
@ChihChengLiang if you are planning the release soon, then I am ok with merging #60 and will stall up the `refactoring` till the new version is released.

/cc @pipermerriam @carver

**Comment by Bhargavasomu** [comment]
@ChihChengLiang another solution would be to remove the file `py.typed` in the next release. I have tested by installing `py_ecc` without the file `py.typed` with the `latest master`. Could you also please confirm that removing `py.typed` would solve the `mypy` issue temporarily?

**Comment by pipermerriam** [comment]
@Bhargavasomu removing `py.typed` isn't an acceptable solution.  We need to fix the typing as soon as possible now that the other PR has been merged.

**Comment by pipermerriam** [comment]
Also @Bhargavasomu want to re-iterate @carver 's point about major version bumping and that we need to address any backwards incompatibilities that may have crept in.

**Comment by ChihChengLiang** [comment]
Alright, don't do `py.typed` removal. @pipermerriam do you prefer any of my proposed options? Especially option 2, it makes BLS and the refactor happy, but not sure if you like it.

**Comment by hwwhww** [comment]
For option 2, IMO we should keep `master` to be the main branch for other PRs.
So, not good-looking but practically, revert `master` branch to 1.4.8, and then see if we can merge #60 without breaking py-evm.

**Comment by Bhargavasomu** [comment]
@pipermerriam the other PR of mine which recently got merged, is only one phase of the 2 phases of refactoring I had in mind. Even if we do `type hinting` now, I think it would still be messy. I'll take care of the `backward incompatibilities`.

@ChihChengLiang could you give a rough time estimate of when you are planning to release the new version with `BLS Signatures`. I ask this so as to know, if I can speed up the process of refactoring and be in par with the schedule of release, so that both the `refactoring` and the `BLS Features` are done in the new release.

**Comment by ChihChengLiang** [comment]
The BLS Signatures is already complete and can be released at any moment.

**Comment by ChihChengLiang** [comment]
The desired release time of BLS Signatures is as soon as possible. But I don't want to rush you for the refactoring. So prefer some way the BLS Signatures can be released soon and you can optimize the refactor as well.

**Comment by carver** [comment]
> But I don't want to rush you for the refactoring.

Yes, rushing through to a new major version is . Even if we did it, we still have maintain the previous major version for a while (duplicating PRs to the old branch).

**Comment by carver** [comment]
Semantic nitpick: a `refactor` is a restructuring of code that [does _not_ change the external API](https://en.wikipedia.org/wiki/Code_refactoring). Up until now, I've been presuming that this was the intention of @Bhargavasomu 's refactor. Maybe we can stretch refactor to mean adding new APIs without breaking old ones.

If the intention is to redesign the API, then let's stop calling it a refactor. It's a rewrite or upgrade.

**Comment by Bhargavasomu** [comment]
This issue can be closed by #56

**Comment by ChihChengLiang** [comment]
fixed in #56
